packages/alfa-aria/src/node.ts

```diff
@@ -222,7 +223,65 @@ export namespace Node {
 
     const root = node.root({ flattened: true });
 
-    build(root, device);
+    // Before we start constructing the accessibility tree, we need to resolve
+    // explicit ownership of elements as specified by the `aria-owns` attribute.
+    // https://w3c.github.io/aria/#aria-owns
+
+    // Find all elements in the tree. As explicit ownership is specified via ID
+    // references, it cannot cross shadow or document boundaries.
+    const elements = root.inclusiveDescendants().filter(dom.Element.isElement);
+
+    // Build a map from ID -> element to allow fast resolution of ID references.
+    // The collected references are added to the map in reverse order to ensure
+    // that the first occurrence of a given ID is what ends up in the map in
+    // event of duplicates.
```

Comment:
I'm wondering whether this map should be made a somewhat global object (probably attached to the `Document`) since there are many case where we need to access elements by ID… Having this available easily could speed things a bit.

Reply:
I've thought about this as well, though I don't yet have a good answer on how to best go about it. I don't want to expose these kinds of lookup tables in public API as that feels like an awfully leaky abstraction. I'd also like to avoid APIs similar to [`Document.getElementById()`](https://developer.mozilla.org/en-US/docs/Web/API/Document/getElementById) as that also feels leaky. I'd much rather we come up with a general query API that uses structured queries that we can then optimise behind the scenes. That is, rather than have N different lookup tables to improve various queries, we have a single query API that we can optimise as needed to ensure a balanced trade-off between time and memory.